applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp

```diff
@@ -0,0 +1,148 @@
+        << "Number of Nodes in Reduced Space: " << num_nodes_reduced_space << " / " << num_of_nodes << std::endl;
+
+    // Construct and initialize correlation matrix
+    BuiltinTimer build_cl_matrix_timer;
+    DenseMatrixType correlation_matrix;
+    correlation_matrix.resize(num_nodes_reduced_space,num_nodes_reduced_space);
+    // Assemble correlation matrix
+    const auto it_node_reduced_begin = reduced_space_nodes.begin();
+    IndexPartition<unsigned int>(num_nodes_reduced_space).for_each(
+        [this, num_nodes_reduced_space, it_node_reduced_begin, &correlation_matrix](unsigned int row_counter){
+            auto it_node = it_node_reduced_begin + row_counter;
+            for( int column_counter = 0; column_counter < num_nodes_reduced_space; column_counter++){
+                auto it_node_2 = it_node_reduced_begin + column_counter;
+                correlation_matrix(row_counter, column_counter) = CorrelationFunction(**it_node, **it_node_2, this->mCorrelationLength);
+            }
+        }
+    );
+
+    KRATOS_INFO_IF("PerturbGeometrySubgridUtility: Build Correlation Matrix Time", mEchoLevel > 0)
+        << build_cl_matrix_timer.ElapsedSeconds() << std::endl;
+
+    // Construct eigensolver and solve eigenproblem
+    DenseVectorType eigenvalues; // Vector is resized inside the solver
+    DenseMatrixType eigenvectors; // Matrix is resized inside the solver
+    DenseMatrixType dummy;
+    mpEigenSolver->Solve(correlation_matrix, dummy, eigenvalues, eigenvectors);
+
+    // Find number of required eigenvalues to statisfy convergence criterion
+    // Eigenvalues are sorted in ascending order and are normalized to an euclidean length of one!!
+    double total_sum_eigenvalues = 0.0;
+    double reduced_sum_eigenvalues = 0.0;
+    int num_eigenvalues_required = 0;
+
+    for( std::size_t i = 0; i < eigenvalues.size(); ++i){
+        total_sum_eigenvalues += eigenvalues(i);
+    }
+    for( std::size_t i = 0; i < eigenvalues.size(); ++i){
+        reduced_sum_eigenvalues += eigenvalues(i);
+        num_eigenvalues_required++;
+        if( reduced_sum_eigenvalues > (1 - mTruncationError)*total_sum_eigenvalues){
+            KRATOS_INFO_IF("PerturbGeometrySubgridUtility", mEchoLevel > 0)
+                << "Truncation Error (" <<  mTruncationError
+                << ") is achieved with " << num_eigenvalues_required << " Eigenvalues" << std::endl;
+            break;
+        }
+    }
+
+    int num_random_variables = num_eigenvalues_required;
+
+    // Get and resize final perturbation matrix
+    DenseMatrixType& rPerturbationMatrix = *mpPerturbationMatrix;
+    rPerturbationMatrix.resize(num_of_nodes, num_random_variables);
+
+    // Construct Correlation vector
+    DenseVectorType correlation_vector;
+    correlation_vector.resize(num_nodes_reduced_space);
+    // Generate random field over full domain
+    BuiltinTimer assemble_random_field_time;
+    #pragma omp parallel for firstprivate(correlation_vector)
```

Comment:
this can also be done using the Parallel Utilities once #7261 is merged

Reply:
Shall I wait until #7261 is merged or make the changes afterwards in a new PR?

Reply:
no need, we can refactor later